### PR TITLE
feat: add plugin-usergroups

### PR DESCRIPTION
**Pull request by songyx**
add new plugin usergroups.
1.add collections usergroups. usergroups m2m users. usergroups o2m usergroups.
2.with a db hook,new users will be add to the default usergroups.
3.need to be plug in before users.

install:
if you have installed.you need insatll again.
yarn nocobase install -f.
it will create new tables:usergroups,usergroupsUsers

### CODE REVIEW

File: packages/app/server/src/config/plugins.ts [hunk 1]
```diff
@@ -7,6 +7,7 @@ export default [
   '@nocobase/plugin-ui-routes-storage',
   '@nocobase/plugin-file-manager',
   '@nocobase/plugin-system-settings',
+  '@nocobase/plugin-usergroups',
```
Comment (chenos) [inline packages/app/server/src/config/plugins.ts:10]: user groups 是两个词，`plugin-user-groups`

File: packages/plugins/usergroups/src/collections/usergroups.ts [hunk 2]
```diff
@@ -0,0 +1,94 @@
+import { CollectionOptions } from '@nocobase/database';
+
+export default {
+    name: 'usergroups',
```
Comment (chenos) [inline packages/plugins/usergroups/src/collections/usergroups.ts:4]: user groups 两个词，userGroups

File: packages/plugins/usergroups/src/collections/usergroups.ts [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+    title: '{{t("UserGroups")}}',
```
Comment (chenos) [inline packages/plugins/usergroups/src/collections/usergroups.ts:5]: 显示文案 `User groups` 分开，groups 不大写

File: packages/plugins/usergroups/src/collections/usergroups.ts [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+            autoIncrement: true,
+            primaryKey: true,
+            allowNull: false,
+            interface: 'id',
+            uiSchema: { type: 'number', title: '{{t("ID")}}', 'x-component': 'InputNumber', 'x-read-pretty': true },
+            // const copy = readonly()
+        },
+        {
+            type: 'string',
+            name: 'name',
```
Comment (chenos) [inline packages/plugins/usergroups/src/collections/usergroups.ts:25]: 这个 name 是标识还是标题？

File: packages/plugins/usergroups/src/collections/usergroups.ts [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+    title: '{{t("UserGroups")}}',
+    sortable: 'sort',
+    model: 'UserGroupModel',
+    filterTargetKey: 'name',
+    createdBy: true,
+    updatedBy: true,
+    logging: true,
+    fields: [
+        {
+            name: 'id',
```
Comment (chenos) [inline packages/plugins/usergroups/src/collections/usergroups.ts:14]: 用 id 关联可能会有迁移问题，可以类似之前其他插件做法用 uid

File: packages/plugins/usergroups/src/collections/usergroups.ts [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+        {//-1 deleted,0 created,1 enable,2 disable
+            type: 'integer',
+            name: 'status',
+            defaultValue: '1',
+        },
+        {//self assocations which will get subgroups
+            interface: 'm2o',
+            type: 'belongsTo',
+            name: 'parent',
+            target: 'usergroups',
```
Comment (chenos) [inline packages/plugins/usergroups/src/collections/usergroups.ts:38]: userGroups

File: packages/plugins/usergroups/src/collections/usergroups.ts [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+            foreignKey: 'pid',
+            uiSchema: {
+                title: '{{t("parentGroup")}}',
```
Comment (chenos) [inline packages/plugins/usergroups/src/collections/usergroups.ts:41]: Parent group

File: packages/plugins/usergroups/src/collections/usergroups.ts [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+            },
+        },
+        {
+            interface: 'o2m',
+            type: 'hasMany',
+            name: 'children',
+            target: 'usergroups',
+            foreignKey: 'pid',
+            uiSchema: {
+                title: '{{t("subGroups")}}',
```
Comment (chenos) [inline packages/plugins/usergroups/src/collections/usergroups.ts:59]: Sub-groups

File: packages/plugins/usergroups/src/collections/usergroups.ts [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+                        value: 'id',
+                    },
+                },
+            },
+        },
+        {
+            interface: 'o2m',
+            type: 'hasMany',
+            name: 'children',
+            target: 'usergroups',
```
Comment (chenos) [inline packages/plugins/usergroups/src/collections/usergroups.ts:56]: userGroups

File: packages/plugins/usergroups/src/collections/usergroups.ts [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+        {
+            interface: 'm2m',
+            type: 'belongsToMany',
+            name: 'users',
+            target: 'users',
+            foreignKey: 'groupId',
+            otherKey: 'userId',
+            sourceKey: 'id',
+            targetKey: 'id',
+            through: 'usergroupsUsers',
```
Comment (chenos) [inline packages/plugins/usergroups/src/collections/usergroups.ts:79]: userGroupsUsers

File: packages/plugins/usergroups/src/collections/usergroups.ts [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+            uiSchema: {
+                type: 'array',
+                title: '{{t("users")}}',
```
Comment (chenos) [inline packages/plugins/usergroups/src/collections/usergroups.ts:82]: Users

File: packages/plugins/usergroups/src/collections/usergroupsUsers.ts [hunk 12]
```diff
@@ -0,0 +1,11 @@
+import { CollectionOptions } from '@nocobase/database';
+
+export default {
+    name: 'usergroupsUsers',
```
Comment (chenos) [inline packages/plugins/usergroups/src/collections/usergroupsUsers.ts:4]: userGroupsUsers

File: packages/plugins/usergroups/src/collections/users.ts [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+      name: 'usergroups',
+      target: 'usergroups',
+      foreignKey: 'userId',
+      otherKey: 'groupId',
+      sourceKey: 'id',
+      targetKey: 'id',
+      through: 'usergroupsUsers',
+      uiSchema: {
+        type: 'array',
+        title: '{{t("userGroups")}}',
```
Comment (chenos) [inline packages/plugins/usergroups/src/collections/users.ts:18]: User groups

File: packages/plugins/usergroups/src/hooks/afterUsersCreate.ts [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+            if (!model.associations) {
+
+                const defaultGroup = await app.db.getRepository('usergroups').findOne({
+                    filter: {
+                        name: 'default',
+                    },
+                });
+
+                //create the default assocations belong to the default usergroup.
+                await model.setUsergroups([defaultGroup], { transaction });
```
Comment (chenos) [inline packages/plugins/usergroups/src/hooks/afterUsersCreate.ts:31]: 如果前端传 groups 了，这里不应该用默认值覆盖

File: packages/plugins/usergroups/src/locale/en-US.ts [hunk 15]
```diff
@@ -0,0 +1,7 @@
+export default {
+  'userGroups': 'userGroups',
```
Comment (chenos) [inline packages/plugins/usergroups/src/locale/en-US.ts:2]: 这里的 key 直接应该是可读的文案，比如 User groups

**Review by chenos — CHANGES_REQUESTED** [review]

File: packages/app/server/src/config/plugins.ts (hunk 1 above)
Comment (songyx) [reply packages/app/server/src/config/plugins.ts:10]: ok，建议将插件命名规则列出。

**Pull request closed without merging**